src/AbstractShop.java

```diff
@@ -0,0 +1,80 @@
+import java.util.ArrayList;
+import java.util.HashMap;
+
+public abstract class AbstractShop implements IShop {
+    /**
+     * hashMap contains products and it's unique id
+     */
+     protected HashMap<Integer,IProduct> productMap = new HashMap<>();
+    /**
+     * static member to remember unique id
+     */
+    protected static Integer count = 1;
+
+    /**
+     *
+     * @param id
+     * @param user
+     * @return product if user can buy otherwise null
+     */
+    @Override
+    public ArrayList<MaterialStack> buyProduct (Integer id,IUser user) {
+        if (!productMap.containsKey(id)) {
+            return null;
+        }
+        boolean checkExist = false;
+        ArrayList<MaterialStack> cost = productMap.get(id).getCost();
+        for (MaterialStack materialStackForUser :user.getUserMaterials()) {

```

Comment:
Double nested for loops are hard to follow, especially if formatting is not consistent.